core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -259,6 +261,209 @@ public static SEXP tryDispatchFromPrimitive(Context context, Environment rho, Fu
     return method.doApply(context, rho, newArgs);
   }
 
+  private static SEXP handleS4object(@Current Context context, SEXP source, String functionName, PairList args,
+                                     Environment rho, int argumentIndex, ArrayList<String> allArgClasses) {
+    // for now the namespace is hardcoded (":base") since all primitives are from the base package. But to
+    // allow this function to also work with standardGeneric the namespace needs to be identified at runtime
+    // since these standardGenerics migh be in .GlobalEnv or elsewhere.
+    String functionEnvName = ".__T__" + functionName + ":base";
+    Environment functionEnv = (Environment) context.getGlobalEnvironment().findVariable(context, Symbol.get(functionEnvName));
+    SEXP function = null;
+    PairList.Builder allArgs = new PairList.Builder();
+    // S4 methods for each generic function is stored in an environment. methods for each signature is stored
+    // separately using the signature as name. for example
+    // setMethod("[", signature("AA","BB","CC"), function(x, i, j, ...))
+    // is stored as `AA#BB#CC` in an environment named `.__T__[:base`
+    // here we get the first method from the method environment and split the name by # to know what the expected
+    // signature length is. This might be longer the length of arguments and #ANY should be used for missing
+    // arguments. In case signature is shorted than the number of arguments we don't need to evaluate the extra
+    // arguments.
+    int signatureLength = ((Symbol) functionEnv.getFrame().getSymbols().toArray()[0]).getPrintName().split("#").length;
+
+    Object[][] allPossibleSignatures = generateAllPossibleSignatures(context, rho, args, null, signatureLength, 0);
+
+    //function = findFunctionMatchingAnyOfSignatures(context, functionEnv, allPossibleSignatures, 0, Symbol.UNBOUND_VALUE);
+
+    Map<Double, SEXP> methods = findFunctionMatchingAnyOfSignatures(context, functionEnv, allPossibleSignatures, 0, null);
+
+    if(methods.isEmpty()) {
+      throw new EvalException("object of type 'S4' is not subsettable");
+    }
+
+    TreeMap<Double, SEXP> map = new TreeMap<>(methods);
+    function = map.get(map.firstKey());
+
+    allArgs.add(source);
+    allArgs.add(args.getElementAsSEXP(1));
+    SEXP result = context.evaluate(new FunctionCall(function, allArgs.build()));
+    return result;
+  }
+
+  private static Map<Double, SEXP> findFunctionMatchingAnyOfSignatures(Context context, Environment functionEnv, Object[][] signatures, int useSignature, Map<Double, SEXP> methods) {
+    // recursively go through arguments and get their class and lookup until the signature is found.
+    // Issues:
+    // - Stops with the first hit
+    // - Is not aware of distance between classes of input arguments to different method signatures
+    if(methods == null) {
+      methods = new HashMap<>(signatures.length);
+    }
+    if(useSignature < signatures.length) {
```

Comment:
Probably would make this code clearer to use a for loop instead of recursion here. Isn't "useSignature" actually "signatureIndex" ? 

```
for(int signatureIndex=0; signatureIndex < signatures.length; ++i) {

}
```